ldi-core/ldes-client/tree-node-supplier/src/main/java/ldes/client/treenodesupplier/repository/filebased/FileBasedMemberRepository.java

```diff
@@ -0,0 +1,57 @@
+package ldes.client.treenodesupplier.repository.filebased;
+
+import ldes.client.treenodesupplier.domain.entities.MemberRecord;
+import ldes.client.treenodesupplier.domain.valueobject.MemberStatus;
+import ldes.client.treenodesupplier.repository.MemberRepository;
+import ldes.client.treenodesupplier.repository.filebased.mapper.MemberRecordMapper;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class FileBasedMemberRepository implements MemberRepository {
+	public static final String UNPROCESSED_MEMBERS = "membersUnprocessed.txt";
+	public static final String PROCESSED_MEMBERS = "membersProcessed.txt";
+
+	private final FileManagerFactory fileManagerFactory = FileManagerFactory.getInstance();
+	private final FileManager fileManager = fileManagerFactory.getFileManager();
+	private final MemberRecordMapper mapper = new MemberRecordMapper();
+
+	@Override
+	public Optional<MemberRecord> getUnprocessedTreeMember() {
+		List<MemberRecord> unprocessed = fileManager.getRecords(UNPROCESSED_MEMBERS).map(mapper::toMemberRecord)
+				.collect(Collectors.toList());
+		if (unprocessed.isEmpty()) {
+			return Optional.empty();
+		}
+		return Optional.of(unprocessed.remove(0));
+	}
+
+	@Override
+	public boolean isProcessed(MemberRecord member) {
+		List<MemberRecord> processed = fileManager.getRecords(PROCESSED_MEMBERS).map(mapper::toMemberRecord).toList();
+		return processed.contains(member);
+	}
+
+	@Override
+	public void saveTreeMember(MemberRecord treeMember) {
+		if (treeMember.getMemberStatus() == MemberStatus.PROCESSED) {
+			fileManager.appendRecord(PROCESSED_MEMBERS, mapper.fromMemberRecord(treeMember));
+			removeMemberFromUnprocessed(treeMember);
+		} else if (treeMember.getMemberStatus() == MemberStatus.UNPROCESSED) {
+			fileManager.appendRecord(UNPROCESSED_MEMBERS, mapper.fromMemberRecord(treeMember));
+		}
+	}
+
+	private void removeMemberFromUnprocessed(MemberRecord treeMember) {
+		List<MemberRecord> unprocessed = fileManager.getRecords(UNPROCESSED_MEMBERS).map(mapper::toMemberRecord)
+				.collect(Collectors.toList());
+		unprocessed.remove(treeMember);
+		fileManager.createNewRecords(UNPROCESSED_MEMBERS, unprocessed.stream().map(mapper::fromMemberRecord));
```

Comment:
What happens if the application crashes here? We already removed the old file?

Reply:
General crashes such as OutOfMemory or halting the application, wouldn't do any harm since the thread would complete its run.